### PR TITLE
Scripts/Spells: Fixed Pet Gargoyle

### PR DESCRIPTION
**Pull request by HoustonCore**
**Changes proposed:**

-  Added a check on the goal of the gargoyle, but I'm not sure if pet should attack a random mob or target a character after the killer mobs.
-  Now pet is coming from above. Before that, pet had been around.
-  Moved the gargoyle strike spell to a separate script.

**Target branch(es):** 3.3.5/master

- [√ ] 3.3.5

**Issues addressed:**

Closes #  https://github.com/TrinityCore/TrinityCore/issues/24682 maybe fix.

**Tests performed:**

Build, tested in-game.
Need more testing and feedback.

**Follow-up:**
Sorry for my English, I'm waiting for constructive criticism, and not here is not so there is not so, I ask for help in solving the standards.

### CODE REVIEW

File: src/server/scripts/Pet/pet_dk.cpp [hunk 1]
```diff
@@ -105,6 +119,109 @@ class npc_pet_dk_ebon_gargoyle : public CreatureScript
...
+                {
+                    if (despawnTimer > diff)
+                        despawnTimer -= diff;
+                    else
+                        me->DespawnOrUnsummon();
+
+                    return;
+                }
+
+                if (spellCastTime <= diff)
```
Comment (Keader) [inline src/server/scripts/Pet/pet_dk.cpp:137]: why not handling it with event map/taskscheduler?

File: src/server/scripts/Pet/pet_dk.cpp [hunk 2]
```diff
@@ -74,9 +68,28 @@ class npc_pet_dk_ebon_gargoyle : public CreatureScript
...
+                {
+                    interrupted = true;
+                    events.ScheduleEvent(EVENT_INTERRUPT_OVER, 20s);
+                }
+            }
+
+            bool CheckCast()
+            {
+                if (me->IsWithinMeleeRange(me->GetVictim()))
+                    return roll_chance_i(40);
```
Comment (Keader) [inline src/server/scripts/Pet/pet_dk.cpp:83]: ???

File: src/server/scripts/Pet/pet_dk.cpp [hunk 3]
```diff
@@ -74,9 +68,28 @@ class npc_pet_dk_ebon_gargoyle : public CreatureScript
                     owner->RemoveAurasDueToSpell(SPELL_DK_SUMMON_GARGOYLE_2);
             }
 
+            void IsInteruptSpell(SpellInfo const* spell)
+            {
+                if (spell->HasEffect(SPELL_EFFECT_INTERRUPT_CAST))
```
Comment (Keader) [inline src/server/scripts/Pet/pet_dk.cpp:73]: why?
SpellHistory already handle with it.

File: src/server/scripts/Pet/pet_dk.cpp [hunk 4]
```diff
@@ -105,6 +119,109 @@ class npc_pet_dk_ebon_gargoyle : public CreatureScript
...
+                    {
+                        case EVENT_INTERRUPT_OVER:
+                            interrupted = false;
+                            break;
+                        default:
+                            break;
+                    }
+                }
+
+                Unit* owner = me->GetOwner();
```
Comment (Keader) [inline src/server/scripts/Pet/pet_dk.cpp:169]: do it inside of UpdateAI break my heart :(

File: src/server/scripts/Pet/pet_dk.cpp [hunk 5]
```diff
@@ -105,6 +119,109 @@ class npc_pet_dk_ebon_gargoyle : public CreatureScript
...
+                        EnableSpellCast = true;
+                }
+                else 
+                    spellCastTime -= diff;
+
+                if (initStatTimer <= diff)
+                {
+                    if (Unit* owner = me->GetOwner())
+                    {
+                        me->SetMaxHealth(uint32(owner->GetMaxHealth() * 0.8f));
```
Comment (Jildor) [inline src/server/scripts/Pet/pet_dk.cpp:149]: I don't think this should go here, also, any proof on that value?

File: src/server/scripts/Pet/pet_dk.cpp [hunk 6]
```diff
@@ -30,41 +30,35 @@
...
+    EVENT_INTERRUPT_OVER = 1
 };
 
 class npc_pet_dk_ebon_gargoyle : public CreatureScript
 {
     public:
         npc_pet_dk_ebon_gargoyle() : CreatureScript("npc_pet_dk_ebon_gargoyle") { }
 
-        struct npc_pet_dk_ebon_gargoyleAI : CasterAI
+        struct npc_pet_dk_ebon_gargoyleAI : ScriptedAI
```
Comment (Jildor) [inline src/server/scripts/Pet/pet_dk.cpp:50]: Shouldn't be casterAI?

**Comment by offl** [comment]
> Changes proposed:
>
>Fixed pet gargoyle
>Fixed summoned pet
>Fixed strike gargoyle

Could you please write more detailed explanation (how exactly those issues are fixed, what is current behavior and expected behavior, which source did you used) so more people can help in reviewing this PR? Because I see tens of undocumented small fixes and some of them looks hacky or just weird

**Comment by jackpoz** [comment]
it helps also if the C++ code has comments in the code itself too

**Comment by HoustonCore** [comment]
> > Changes proposed:
> > Fixed pet gargoyle
> > Fixed summoned pet
> > Fixed strike gargoyle
> 
> Could you please write more detailed explanation (how exactly those issues are fixed, what is current behavior and expected behavior, which source did you used) so more people can help in reviewing this PR? Because I see tens of undocumented small fixes and some of them looks hacky or just weird

 - Fixed pet gargoyle -

**Comment by HoustonCore** [comment]
> > Changes proposed:
> > Fixed pet gargoyle
> > Fixed summoned pet
> > Fixed strike gargoyle
> 
> Could you please write more detailed explanation (how exactly those issues are fixed, what is current behavior and expected behavior, which source did you used) so more people can help in reviewing this PR? Because I see tens of undocumented small fixes and some of them looks hacky or just weird

Done

File: src/server/scripts/Pet/pet_dk.cpp (hunk 2 above)
Comment (HoustonCore) [reply src/server/scripts/Pet/pet_dk.cpp:83]: According to the logic of Peta needs to choose any target, after the killing of a mob.

File: src/server/scripts/Pet/pet_dk.cpp (hunk 3 above)
Comment (HoustonCore) [reply src/server/scripts/Pet/pet_dk.cpp:73]: This script 2015 year, i say help me move new standart.

File: src/server/scripts/Pet/pet_dk.cpp (hunk 1 above)
Comment (HoustonCore) [reply src/server/scripts/Pet/pet_dk.cpp:137]: Okey i move event.

File: src/server/scripts/Pet/pet_dk.cpp (hunk 5 above)
Comment (HoustonCore) [reply src/server/scripts/Pet/pet_dk.cpp:149]: Я поищю данную тему о хп гаргульи, пока удалю.

Comment (HoustonCore) [reply src/server/scripts/Pet/pet_dk.cpp:149]: I'll look for this topic about HP gargoyles until I delete it.

File: src/server/scripts/Pet/pet_dk.cpp (hunk 6 above)
Comment (HoustonCore) [reply src/server/scripts/Pet/pet_dk.cpp:50]: It's work fine in ScriptedAI.

**Comment by HoustonCore** [comment]
I will close, there is no time to adapt.

**Pull request closed without merging**